qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,529 @@
+
+        if len(args1) == 0:
+            raise EvalProxyError("Nothing is passed to check the range")
+
+        return self._time_range(args1, args2, gmt)
+
+
+class PACResolver(object):
+    """Evaluate PAC script files and resolve proxies."""
+
+    @staticmethod
+    def _parse_proxy_entry(proxy_str):
+        """Parse one proxy string entry, as described in PAC specification."""
+        config = [c for c in proxy_str.split(' ') if c != ""]
+        if len(config) == 0:
+            raise ParseProxyError("Empty proxy entry")
+        try:
+            if config[0] == "DIRECT":
+                if len(config) != 1:
+                    raise ParseProxyError("Invalid number of parameters")
+                return QNetworkProxy(QNetworkProxy.NoProxy)
+            elif config[0] == "PROXY":
+                if len(config) != 2:
+                    raise ParseProxyError("Invalid number of parameters")
+                host, _, port = config[1].partition(':')
+                return QNetworkProxy(QNetworkProxy.HttpProxy, host, int(port))
+            elif config[0] == "SOCKS":
+                if len(config) != 2:
+                    raise ParseProxyError("Invalid number of parameters")
+                host, _, port = config[1].partition(':')
+                return QNetworkProxy(QNetworkProxy.Socks5Proxy, host,
+                                     int(port))
+            else:
+                err = "Unknown proxy type: {}"
+                raise ParseProxyError(err.format(config[0]))
+        except ValueError:
+            raise ParseProxyError("Invalid port number")
+
+    @staticmethod
+    def _parse_proxy_string(proxy_str):
+        proxies = proxy_str.split(';')
+        return [PACResolver._parse_proxy_entry(x) for x in proxies]
+
+    def __init__(self, pac_str, fixed_dt=None):
+        """Create a PAC resolver.
+
+        Args:
+            pac_str: JavaScript code containing PAC resolver.
+            fixed_dt: Assume fixed date and time instead of system time.
+        """
+        self._engine = QJSEngine()
+
+        self._ctx = _PACContext(self._engine, fixed_dt)
+        self._engine.globalObject().setProperty("PAC",
+            self._engine.newQObject(self._ctx))
+        proxy_config = self._engine.newObject()
+        proxy_config.setProperty("bindings", self._engine.newObject())
+        self._engine.globalObject().setProperty("ProxyConfig", proxy_config)
+        ctx_meta = self._ctx.metaObject()
+        for i in range(ctx_meta.methodCount()):
+            m = ctx_meta.method(i)
+            if m.typeName() == "QJSValue":
+                if m.parameterCount() == 1 and \
+                   m.parameterType(0) == QMetaType.QVariant:
+                    call_str = "PAC.{0}([].slice.call(arguments))"
+                else:
+                    call_str = "PAC.{0}.apply(PAC, arguments)"
+                decl_str = '''
+                    function {0}() {{
+                        var res = ''' + call_str + ''';
```

Comment:
Hm, I think I feared some quotation issues but I see none now. I also don't need to use explicit argument numbers (`{0}`) there too.